lib/cea/cea_utils.js

```diff
@@ -0,0 +1,258 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+goog.provide('shaka.cea.CeaUtils');
+goog.provide('shaka.cea.CeaUtils.StyledChar');
+
+shaka.cea.CeaUtils = class {
+  /**
+   * Emits a closed caption based on the state of the buffer.
+   * @param {!shaka.text.Cue} topLevelCue
+   * @param {!string} stream
+   * @param {!Array<!Array<?shaka.cea.CeaUtils.StyledChar>>} memory
+   * @param {!number} startTime Start time of the cue.
+   * @param {!number} endTime End time of the cue.
+   * @return {?shaka.cea.ICaptionDecoder.ClosedCaption}
+   */
+  static getParsedCaption(topLevelCue, stream, memory, startTime, endTime) {
+    if (startTime >= endTime) {
+      return null;
+    }
+
+    // Find the first and last row that contains characters.
+    let firstNonEmptyRow = -1;
+    let lastNonEmptyRow = -1;
+
+    for (let i = 0; i < memory.length; i++) {
+      if (memory[i].some((e) => e != null && e.getChar().trim() != '')) {
+        firstNonEmptyRow = i;
+        break;
+      }
+    }
+
+    for (let i = memory.length - 1; i >= 0; i--) {
+      if (memory[i].some((e) => e != null && e.getChar().trim() != '')) {
+        lastNonEmptyRow = i;
+        break;
+      }
+    }
+
+    // Exit early if no non-empty row was found.
+    if (firstNonEmptyRow === -1 || lastNonEmptyRow === -1) {
+      return null;
+    }
+
+    // Keeps track of the current styles for a cue being emitted.
+    let currentUnderline = false;
+    let currentItalics = false;
+    let currentTextColor = shaka.cea.CeaUtils.DEFAULT_TXT_COLOR;
+    let currentBackgroundColor = shaka.cea.CeaUtils.DEFAULT_BG_COLOR;
+
+    // Create first cue that will be nested in top level cue. Default styles.
+    let currentCue = shaka.cea.CeaUtils.createStyledCue(
+        startTime, endTime, currentUnderline, currentItalics,
+        currentTextColor, currentBackgroundColor);
+
+    // Logic: Reduce rows into a single top level cue containing nested cues.
+    // Each nested cue corresponds either a style change or a line break.
+
+    for (let i = firstNonEmptyRow; i <= lastNonEmptyRow; i++) {
+      // Find the first and last non-empty characters in this row. We do this so
+      // no styles creep in before/after the first and last non-empty chars.
```

Comment:
This is extra logic that I added to both the CEA-608 and CEA-708 emit. It gets rid of leading and trailing whitespaces having styling, such as underline/italics/color.

This doesn't affect time complexity or efficiency, since we're still processing the same amount of data, just getting the first and last non-empty columns before doing so.